ruby/ql/lib/codeql/ruby/ast/Method.qll

```diff
@@ -140,28 +152,41 @@ class Method extends MethodBase, TMethod {
    * ```
    */
   override predicate isPrivate() {
-    exists(Namespace n, string name |
-      any(Private p).modifiesMethod(n, name) and
-      isDeclaredIn(this, n, name)
-    )
+    this.getVisibilityModifier().getVisibility() = "private"
     or
     // Top-level methods are private members of the Object class
     this.getEnclosingModule() instanceof Toplevel
   }
 
+  override predicate isPublic() { super.isPublic() and not this.isPrivate() }
+
   final override Parameter getParameter(int n) {
     toGenerated(result) = g.getParameters().getChild(n)
   }
 
   final override string toString() { result = this.getName() }
-}
 
-/**
- * Holds if the method `m` has name `name` and is declared in namespace `n`.
- */
-pragma[noinline]
-private predicate isDeclaredIn(MethodBase m, Namespace n, string name) {
-  n = m.getEnclosingModule() and name = m.getName()
+  override VisibilityModifier getVisibilityModifier() {
+    result.getEnclosingModule() = this.getEnclosingModule() and
+    (
+      result.getMethodArgument() = this
+      or
+      result.getMethodArgument().getConstantValue().getStringlikeValue() = this.getName()
+    )
+    or
+    exists(Namespace n, int methodPos | n.getStmt(methodPos) = this |
```

Comment:
Should there be a check that this method does not have an explicit overriding modifier?